### PR TITLE
[EMCAL-565]: Add calibrator structure.

### PR DESCRIPTION
**Pull request by hjbossi**
- Work in progress to add the calibrator structure.

### CODE REVIEW

**Comment by mfasDa** [comment]
@hjbossi There are still merge conflicts that need to be resolved before testing - did you rebase to dev before pushing?

**Comment by hjbossi** [comment]
Hi @mfasDa - the merge conflicts should be resolved. Let me know if you have any other comments!

File: Common/Utils/CMakeLists.txt [hunk 1]
```diff
@@ -38,8 +39,14 @@ o2_target_root_dictionary(CommonUtils
                                   include/CommonUtils/ConfigurationMacroHelper.h
                                   include/CommonUtils/RootSerializableKeyValueStore.h
                                   include/CommonUtils/KeyValParam.h
+<<<<<<< HEAD
                                   include/CommonUtils/VerbosityConfig.h
                                   include/CommonUtils/FileFetcher.h)
+=======
```
Comment (mfasDa) [inline Common/Utils/CMakeLists.txt]: Merge conflicts not resolved.

File: dpl-config.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                "0.40000000000000002"
+            ]
+        },
+        "jetType": "0",
+        "processParticleLevel": "false",
+        "processData": "true"
+    },
+    "internal-dpl-aod-global-analysis-file-sink": "",
+    "internal-dpl-aod-writer": ""
+}
```
Comment (mfasDa) [inline dpl-config.json]: File must not enter at all

File: Detectors/EMCAL/workflow/src/RecoWorkflow.cxx [hunk 3]
```diff
@@ -61,19 +61,22 @@ o2::framework::WorkflowSpec getWorkflow(bool propagateMC,
...
     {"clusters", OutputType::Clusters},
-    {"analysisclusters", OutputType::AnalysisClusters}};
+    {"analysisclusters", OutputType::AnalysisClusters},
+    {"badchannelmap", OutputType::BadChannelMap}};
 
   std::unordered_map<InputType, std::vector<OutputType>> allowedIO;
   allowedIO[InputType::Digits] = std::vector<OutputType>{OutputType::Cells, OutputType::Clusters, OutputType::AnalysisClusters};
   allowedIO[InputType::Cells] = std::vector<OutputType>{OutputType::Cells, OutputType::Clusters, OutputType::AnalysisClusters};
   allowedIO[InputType::Raw] = std::vector<OutputType>{OutputType::Cells};
+  allowedIO[InputType::BoostHisto] = std::vector<OutputType>{OutputType::BadChannelMap};
```
Comment (mfasDa) [inline Detectors/EMCAL/workflow/src/RecoWorkflow.cxx]: Not needed in reco workflow, the calibration params are handled via the internal CCDB interface. I guess you mixed up something with the calibrator workflow.

File: Detectors/EMCAL/workflow/include/EMCALWorkflow/RecoWorkflow.h [hunk 4]
```diff
@@ -28,20 +28,22 @@ namespace reco_workflow
...
-                         Cells,           ///< EMCAL cells
-                         Raw,             ///< EMCAL raw data
-                         Clusters,        ///< EMCAL clusters
-                         AnalysisClusters ///< EMCAL analysis clusters
+enum struct OutputType { Digits,           ///< EMCAL digits
+                         Cells,            ///< EMCAL cells
+                         Raw,              ///< EMCAL raw data
+                         Clusters,         ///< EMCAL clusters
+                         AnalysisClusters, ///< EMCAL analysis clusters
+                         BadChannelMap     ///< EMCAL bad channel map
```
Comment (mfasDa) [inline Detectors/EMCAL/workflow/include/EMCALWorkflow/RecoWorkflow.h]: Not needed in reco workflow, the calibration params are handled via the internal CCDB interface. I guess you mixed up something with the calibrator workflow.

**Review by mfasDa — CHANGES_REQUESTED** [review]
Still not ready for merge. There are still leftovers from the manual merge + there are files added that should not be added. Before merging please rebase again to dev in order to incorporate the changes from @jokonig .

File: Detectors/EMCAL/calibration/src/EMCALChannelData.cxx [hunk 5]
```diff
@@ -83,5 +84,174 @@ bool EMCALChannelData::hasEnoughData() const
   return enough;
 }
 
+//_____________________________________________
+void EMCALChannelData::buildHitAndEnergyMean(double emin, double emax)
```
Comment (mfasDa) [inline Detectors/EMCAL/calibration/src/EMCALChannelData.cxx]: Needs to move to the CalibExtractor, mEsumHisto should no longer be a class member but a return object.

File: Detectors/EMCAL/calibration/src/EMCALChannelData.cxx [hunk 6]
```diff
@@ -83,5 +84,174 @@ bool EMCALChannelData::hasEnoughData() const
...
+// This function builds the scaled hit distribution
+// It normalizes the hits/cell to the mean value of the row and the column of the cell
+// this is done in an iterative procedure (about 5 iterations are needed)
+// with this procedure, the effects of the TRD and SM structures on the EMCal can be minimized
+// The output is a  boost histogram with the hits/cell as a function of cell ID.
+
+/// \param emin -- min. energy for cell amplitudes
+/// \param emax -- max. energy for cell amplitudes
+// ------------------------------------------------------------------------------------------
+void EMCALChannelData::buildHitAndEnergyMeanScaled(double emin, double emax)
```
Comment (mfasDa) [inline Detectors/EMCAL/calibration/src/EMCALChannelData.cxx]: Also need to go to the CalibExtractor.

File: Detectors/EMCAL/calibration/src/EMCALChannelData.cxx [hunk 7]
```diff
@@ -83,5 +84,174 @@ bool EMCALChannelData::hasEnoughData() const
...
+      auto energyAxis = boost::histogram::axis::regular<>(100, 0, 100, "t-texp");
+      Double_t E = energyAxis.value(j);
+      Double_t N = hEnergyScaled.at(j, cell);
+      if (E < emin || E > emax)
+        continue;
+      Esum += E * N;
+      Nsum += N;
+    }
+    if (Nsum > 0.)
+      mEsumHistoScaled(cell, Esum / (Nsum)); //..average energy per hit
```
Comment (mfasDa) [inline Detectors/EMCAL/calibration/src/EMCALChannelData.cxx]: No class member but local variable which serves as return value

File: Detectors/EMCAL/calibration/src/EMCALSlotCalibLinkDef.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+#ifdef __CLING__
+
+#pragma link off all globals;
+#pragma link off all classes;
+#pragma link off all functions;
+
+#pragma link C++ class o2::emcal::EMCALSlotCalib + ;
+
+#endif
```
Comment (mfasDa) [inline Detectors/EMCAL/calibration/src/EMCALSlotCalibLinkDef.h]: Needs to go as well.

**Review by mfasDa — CHANGES_REQUESTED** [review]

**Review by mfasDa — DISMISSED** [review]
OK now from my side

**Comment by mfasDa** [comment]
Should be fine now. @sawenzel @shahor02 Can you please merge when the tests pass?

**Review by mfasDa — APPROVED** [review]

**Comment by mfasDa** [comment]
@sawenzel Do you think this could be merged now?

**Pull request merged**